### PR TITLE
Python: Fix import of refined variable

### PR DESCRIPTION
**Pull request by RasmusWL**
Such as importing Foo from the following file
```py
class Foo: pass
Foo.bar = 42
```

The original approach wasn't quite on point, which I discovered with the tests now added in fb425b73fcac5d88c97a35b0a87479f1613308ff -- so a force push and a lot of rework of the approach later, I think we're at a good enough place now (if performance is reasonable :flushed:)

_(there could still be more work on import resolution, but I'm going to refrain from digging more into it now)_

### CODE REVIEW

File: python/ql/lib/semmle/python/dataflow/new/internal/ImportResolution.qll [hunk 1]
```diff
@@ -64,6 +64,19 @@ private import semmle.python.dataflow.new.internal.DataFlowPrivate
...
+  /**
+   * Holds if there is an ESSA step from `defFrom` to `defTo`, which should be allowed
+   * for import resolution.
+   */
+  private predicate allowedEssaImportStep(EssaDefinition defFrom, EssaDefinition defTo) {
+    // to handle definitions guarded by if-then-else
+    defFrom = defTo.(PhiFunction).getAnInput()
+    or
+    // refined variable
+    // example: https://github.com/nvbn/thefuck/blob/ceeaeab94b5df5a4fe9d94d61e4f6b0bbea96378/thefuck/utils.py#L25-L45
```
Comment (tausbn) [inline python/ql/lib/semmle/python/dataflow/new/internal/ImportResolution.qll:76]: That's a... Colourful project name.

**Review by tausbn — APPROVED** [review]
Looks good to me. This one might have an impact on performance, so I'll eagerly await the results. 🙂

File: python/ql/lib/semmle/python/dataflow/new/internal/ImportResolution.qll (hunk 1 above)
Comment (calumgrant) [reply python/ql/lib/semmle/python/dataflow/new/internal/ImportResolution.qll:76]: I wonder if `Foo` is kinda-colourful as well. foo/bar = [FUBAR](https://dictionary.cambridge.org/dictionary/english/fubar)

**Comment by RasmusWL** [comment]
Tests should pass after a rerun, were only failing due to an unupdated .expected file on `main`

**Review by tausbn — DISMISSED** [review]
Looks good to me. 👍

**Comment by RasmusWL** [comment]
Performance is fine :heavy_check_mark: new results looks ok from what I can tell.

**Review by tausbn — APPROVED** [review]
Nice work! 👍

**Pull request merged**